### ISSUE TITLE
Sync Server issues

### ISSUE THREAD

**Issue by KodySEC** [MISP/MISP#633]
Any ideas?

I get these messages when I try to pull data from another MISP instance.

stream_socket_client(): Unable to set verify locations `/var/www/MISP/app/files/certs/4.pem' `(null)' stream_socket_client(): failed to create an SSL handle stream_socket_client(): Failed to enable crypto stream_socket_client(): unable to connect to ssl://MISP.MISP.MISP:443 (Unknown error) 

stream_socket_client(): SSL operation failed with code 1. OpenSSL Error messages: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed stream_socket_client(): Failed to enable crypto stream_socket_client(): unable to connect to ssl://MISP.MISP.MISP:443 (Unknown error) 

stream_socket_client(): Unable to set verify locations `/var/www/MISP/app/files/certs/3.pem' `(null)' stream_socket_client(): failed to create an SSL handle stream_socket_client(): Failed to enable crypto stream_socket_client(): unable to connect to ssl://MISP.MISP.MISP:443 (Unknown error)

**Comment by iglocska** [comment]
That is indeed weird, could you make sure that the certificate files exist and are readable by MISP? (in /var/www/MISP/app/files/certs/)

**Comment by KodySEC** [comment]
I am going to ask a very noob question. Pride aside here it is.... If I am having two MISP instances/servers share data, do I need to share a cert between the two, and if so, how do I go about generating said cert? I ask because my misp instances share data between each user, however, it will not PUSH/PULL data to another synced server/my community.

**Comment by KodySEC** [comment]
Yes the .pem cert is in /var/www/MISP/app/files/certs/

**Comment by iglocska** [comment]
It basically works like this:

When connecting two instances, (MISP A and MISP B), both MISP A and MISP B need to exchange 2 things: SSL certificates and sync user API keys. The SSL cert allows MISP to verify whether it is connecting to the legitimate instance and the API key is needed to control who can connect to the instance and what privileges they'd have.

1. SSL cert

You should have an SSL certificate that you use for each instance (if you use https, if it's 2 connected non internet facing instances then you could also use http I guess, in which case you don't need this), which can either be signed by a signing authority or it can be self-signed. 

If it is signed by a known signing authority (one that is included in the mozilla cafile) then you don't have to do anything, don't upload any files while creating the connection between your instances, MISP will already be able to validate the certificate of the other instance!

If it is a self-signed certificate, you need export certificate (as a .pem file) of the other MISP instance and import it while setting up the connection. 

The tricky scenario here is that you might have your own signing authority, in which case you need the entire certificate chain included in the .pem file, so you'll have to create a merged file for the certs. 

2. API key

Each MISP user has an API that is inactive unless a special role is assigned to them. By default, you should have a role called "Sync user", this is what you can use to interconnect MISP instances.

Basically, the admin of MISP A will create a user for the owner organisation of MISP B, assign the sync user role to it and send the API key out of bound to the admin of MISP B. MISP B admin will then use this Authentication key when setting up the connection to MISP A. MISP B then does the exact same thing (creating a user for the owner org of MISP A and sending the admin of MISP A the credentials). 

Once all of this is done, you can start exchanging data.

**Comment by KodySEC** [comment]
So I converted the crt that I created when I was setting MISP up for SSL, to a .pem. I exported the new .pem to my local PC, remove and setup a new "Sync server", gave it the base url, ORG name, auth key, and selected the new .pem file. I'm still receiving the following error, and still not able to push/pull data. I should mention that I am able to ping both servers from each other.

stream_socket_client(): Peer certificate CN=`ip-misp.misp.misp' did not match expected CN=`misp.msp.misp.misp ' stream_socket_client(): Failed to enable crypto stream_socket_client(): unable to connect to ssl://misp.misp.misp.misp:443 (Unknown error) 

I really do appreciate all your help. I totally understand if your face is in the palm of your hand and you are shaking your head at me.

**Comment by iglocska** [comment]
Don't worry, these things can be frustrating! Looking at the error message,
it looks like the certificate is defined for ip-misp.misp.misp and the
actual address that you are trying to reach is misp.msp.misp.misp, which
causes the verification to (correctly) fail. Make sure that the certificate
is created for the CN that you will be using for your MISP installation (so
in this case it would be misp.msp.misp.misp). (Also if that's a copy paste,
note the second "msp" instead of misp, just a heads-up!)

**Comment by TrustedForce** [comment]
I really appreciate all your help. With the information you provided me I was able to figure it out. Thanks again.

-Cheers

**Comment by iglocska** [comment]
Excellent, glad it worked out!

**Comment by criptk** [comment]
Hey guys,
I am having a similar problem.
My setup is a little different as the MISP instances aren't directly reachable (one is behind NAT, the one initiating the push/pulls.)
Would this setup be possible ?
I have trying add both server and client certificates as PEM (converted the CRT from the browser export).
but getting this error
-==-=-=---
###  (Error: Connection test failed. Reason: "stream_socket_client(): SSL operation failed with code 1. OpenSSL Error messages:\nerror:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed\nstream_socket_client(): Failed to enable crypto\nstream_socket_client(): unable to connect to ssl:\/\/misp.xxxxx.com:443 (Unknown error)" ) ###

This is from the one initiating the sync from the servers page, the user is created on the remote instance and i have created and used the API Key.

Any insights into where to get more information or how to further debug are welcome =)

Thanks
